ncm-spma/src/main/perl/spma/yumng.pm

```diff
@@ -0,0 +1,810 @@
+    my $pkgs               = $config->getElement(PKGS_TREE)->getTree();
+    my $wanted_pkgs        = Set::Scalar->new();
+    my $wanted_pkgs_locked = Set::Scalar->new();
+    my $found_spma         = 0;
+    my @pkl;
+    my @pkl_v;
+    my @pkl_a;
+
+    for my $name ( keys %$pkgs ) {
+        if ( !$found_spma && substr( unescape $name, 0, 8 ) eq 'ncm-spma' ) {
+            $found_spma = 1;
+        }
+        my $vra = $pkgs->{$name};
+        while ( my ( $vers, $a ) = each(%$vra) ) {
+            my $arches = $a->{arch};
+            if ( exists( $a->{repository} ) ) {
+                foreach my $arch (@$arches) {
+                    if ( $vers ne '_' ) {
+                        push ( @pkl_v, ( unescape $name) . ';' . ( unescape $vers) . '.' . $arch );
+                    } else {
+                        if ( $arch eq '_' ) {
+                            push ( @pkl, unescape $name . ';' );
+                        } else {
+                            push ( @pkl_a, unescape $name . ';' . $arch );
+                        }
+                    }
+                }
+            } else {
+                foreach my $arch ( keys %$arches ) {
+                    if ( $vers ne '_' ) {
+                        push ( @pkl_v, ( unescape $name) . ';' . ( unescape $vers) . '.' . $arch );
+                    } else {
+                        if ( $arch eq '_' ) {
+                            push ( @pkl, unescape $name . ';' );
+                        } else {
+                            push ( @pkl_a, unescape $name . ';' . $arch );
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    if ( !$found_spma ) {
+        $self->error('Package ncm-spma is not present among required packages.');
+        return 0;
+    }
+
+    my $wanted_pkgs_uv = Set::Scalar->new(@pkl);
+    my $wanted_pkgs_v  = Set::Scalar->new(@pkl_v);
+    my $wanted_pkgs_a  = Set::Scalar->new(@pkl_a);
+    while ( defined( my $p = $wanted_pkgs_uv->each ) ) {
+        while ( defined( my $p2 = $wanted_pkgs_v->each ) ) {
+            if ( index( $p2, $p ) == 0 ) {
+                my $pkg = $p;
+                chop($pkg);
+                my $pkg_locked = $p2;
+                $pkg_locked =~ tr/;/-/;
+                $self->info("preferring version-locked $pkg_locked over unlocked $pkg");
+                $wanted_pkgs_uv->delete($p);
+            }
+        }
+    }
+    while ( defined( my $p = $wanted_pkgs_uv->each ) ) {
+        chop($p);
+        $wanted_pkgs->insert($p);
+    }
+    while ( defined( my $p = $wanted_pkgs_v->each ) ) {
+        $p =~ s/;/-/;
+        $wanted_pkgs_locked->insert($p);
+    }
+    while ( defined( my $p = $wanted_pkgs_a->each ) ) {
+        $p =~ s/;/./;
+        $wanted_pkgs->insert($p);
+    }
+
+    if ( !$NoAction ) {
+        # Import GPG keys
+        foreach my $file ( glob "/etc/pki/rpm-gpg/RPM-GPG-KEY*" ) {
```

Comment:
We use GPG signed repositories and force YUM to verify signatures. These files are installed vya redhat-release for OS and our custom package shipping other GPG keys which we control.